cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/proto/ReportGenerationRequest.java

```diff
@@ -0,0 +1,330 @@
+   */
+  public Long getStart() {
+    return start;
+  }
+
+  /**
+   * @return the end of the time range within which the report is generated. All program runs in the report must be
+   *         still start before {@code end}.
+   */
+  public Long getEnd() {
+    return end;
+  }
+
+  /**
+   * @return names of the fields to be included in the final report. Must be valid fields from {@link ReportField}.
+   */
+  public List<String> getFields() {
+    return fields;
+  }
+
+  /**
+   * @return the field to sort the report by. Currently only support a single field.
+   */
+  @Nullable
+  public List<Sort> getSort() {
+    return sort;
+  }
+
+  /**
+   * @return the filters that must be satisfied for every record in the report.
+   */
+  @Nullable
+  public List<Filter> getFilters() {
+    return filters;
+  }
+
+  /**
+   * Validates this {@link ReportGenerationRequest}
+   *
+   * @throws IllegalArgumentException if this request is not valid.
+   */
+  public void validate() {
+    if (start == null) {
+      throw new IllegalArgumentException("'start' must be specified.");
+    }
+    if (end == null) {
+      throw new IllegalArgumentException("'end' must be specified.");
+    }
+    if (fields == null || fields.isEmpty()) {
+      throw new IllegalArgumentException("'fields' must be specified.");
+    }
+    for (String field : fields) {
+      if (!ReportField.isValidField(field)) {
+        throw new IllegalArgumentException(String.format("Invalid field name '%s' in fields.", field));
+      }
+    }
+    // No need to check whether filter field name and type are valid since this is done during JSON deserialization
+    if (filters != null) {
+      Set<String> existingFilterFields = new HashSet<>();
+      for (Filter filterField : filters) {
+        if (existingFilterFields.contains(filterField.getFieldName())) {
+          throw new IllegalArgumentException(String.format("Field '%s' is duplicated in filters.",
+                                                           filterField.getFieldName()));
+        }
+        existingFilterFields.add(filterField.getFieldName());
+      }
+    }
+    if (sort != null) {
+      if (sort.size() > 1) {
+        throw new IllegalArgumentException("Currently only one field is supported in sort.");
+      }
+      for (Sort sortField : sort) {
+        ReportField sortFieldType = ReportField.valueOfFieldName(sortField.getFieldName());
+        if (sortFieldType == null) {
+          throw new IllegalArgumentException(String.format("Invalid field name '%s' in sort.",
+                                                           sortField.getFieldName()));
+        }
+        if (!sortFieldType.isSortable()) {
+          throw new IllegalArgumentException(String.format("Field '%s' in sort is not sortable.",
+                                                           sortField.getFieldName()));
+        }
+      }
+    }
```

Comment:
i think instead of throwing illegalArgumentException immediately, if we can group all the issues with the request in a list and send that as as an exception (if list size is > 0), users can correct all their issues together before submitting the next request.